### ISSUE TITLE
cufflinks dependency problem

### ISSUE THREAD

**Issue by tiagoantao** [galaxyproject/tools-devteam#452]
I just installed a brand new galaxy server (16.10). I installed cufflinks from scratch.

I am having a bug when running the code:

```
Fatal error: Exit code 1 ()
Traceback (most recent call last):
  File "/shed_tools/toolshed.g2.bx.psu.edu/repos/devteam/cufflinks/83bec71c5c9f/cufflinks/cufflinks_wrapper.py", line 10, in <module>
    from galaxy.datatypes.util.gff_util import gff_attributes_to_str, parse_gff_attributes
  File "/galaxy-central/lib/galaxy/datatypes/util/gff_util.py", line 6, in <module>
    from bx.intervals.io import GenomicInterval, MissingFieldError, NiceReaderWrapper, ParseError, GenomicIntervalReader
ImportError: No module named 'bx'`

```

Now, I have tracked the dependency definition and bx is included and seems perfectly correct.

It seems that the wrapper is calling a conda-based python, which has a

```
/export/tool_deps/_conda/envs/__bx-python@0.7.3/bin/python
```

inside.

But for some reason the bx package is not being made available in the environment being run by cufflinks.

This is a brand new installation and I suspect that there are no strange interactions or ugly hacks at this stage.

**Comment by nsoranzo** [comment]
I suspect that's a problem due to using miniconda2 (default in 16.10) instead of miniconda3 (default in 17.01). Ping @mvdbeek

**Comment by tiagoantao** [comment]
I upgraded to 17.01 and I have exactly the sample problem again.

**Comment by nsoranzo** [comment]
Upgrading to the new release cannot upgrade your Conda installation unfortunately (if you already have miniconda2), because that would break all Conda dependencies already installed. Or is this a fresh install?

**Comment by tiagoantao** [comment]
Fresh install, Docker based. I decided to start from the very beginning as our previous installation was over-hacked.

**Comment by mvdbeek** [comment]
Cufflinks is one of the tools that needs Galaxy's python currently, because it imports from the datatypes. You shouldn't get the same error on 17.01 and 16.10 though. This tool can currently be used only with toolshed package dependencies, are they installed? We'll have to upgrade this to support conda and remove the import from galaxy's codebase.

**Comment by nsoranzo** [comment]
> Cufflinks is one of the tools that needs Galaxy's python currently, because it imports from the datatypes.

True, in fact it is in https://github.com/jmchilton/galaxy/blob/eb6c0e410e600171fa3ce4fffb78bc205aad5292/lib/galaxy/tools/__init__.py#L137

@mvdbeek cufflinks tool is Conda-ready since https://github.com/galaxyproject/tools-devteam/pull/412, the test passes:

https://travis-ci.org/galaxyproject/tools-devteam/jobs/199322959#L2195

**Comment by tiagoantao** [comment]
I think I have found the cause:
bx is installed in the python 2 environment
but cufflinks is run on python 3. so the library is not found.

Note again that this happens on a stock galaxy (docker), not tweaks at all.

**Comment by bgruening** [comment]
@tiagoantao is this 17.01?
I guess we need to explicitly put python2.7 into the tool requirements.

**Comment by tiagoantao** [comment]
I inspected the python lib created for the job and a python2.7 directory is there (with bx), but it is not added to the sys.path.

Which I think it **makes sense** because the code is being executed in python 3.

**Comment by tiagoantao** [comment]
I do not have a clue about the best practices here, but it seems that cufflinks needs to depend on a bx-python3 or something like that, not the standard bx which is for python 2

**Comment by mvdbeek** [comment]
IMO the right thing to do is to not import from galaxy. If no one beats me
to it I can have a look at the options we have.

**Comment by tiagoantao** [comment]
If it helps: I did a bit more research and it seems bx-python is python 2 ONLY.
What I did was to change the wrapper (cufflinks_wrapper.xml) to call python2

**Comment by nsoranzo** [comment]
bx-python is indeed Python2 only, so the Conda environment containing the 3 requirements should contain Python2. We need the Galaxy logs to see what's going on.

**Comment by tiagoantao** [comment]
Do you want me to send anything (log-wise)?

**Comment by nsoranzo** [comment]
Yes, can you post the Galaxy handler logs of a cufflinks tool execution, please?

**Comment by tiagoantao** [comment]
```
galaxy.jobs DEBUG 2017-02-17 22:55:49,748 (42) Working directory for job is: /export/galaxy-central/database/job_working_directory/000/42
galaxy.jobs.handler DEBUG 2017-02-17 22:55:49,754 (42) Dispatching to slurm runner
galaxy.jobs DEBUG 2017-02-17 22:55:49,783 (42) Persisting job destination (destination id: slurm_cluster)
galaxy.jobs.runners DEBUG 2017-02-17 22:55:49,793 Job [42] queued (39.228 ms)
galaxy.jobs.handler INFO 2017-02-17 22:55:49,805 (42) Job dispatched
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,979 Find dependency bx-python version 0.7.3
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,979 Using dependency bx-python version 0.7.3 of type conda
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,979 Find dependency six version 1.9.0
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,980 Using dependency six version 1.9.0 of type conda
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,980 Find dependency cufflinks version 2.2.1
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,982 Using dependency cufflinks version 2.2.1 of type tool_shed_package
galaxy.tools.deps.conda_util DEBUG 2017-02-17 22:55:49,982 Executing command: /export/tool_deps/_conda/bin/conda list --name __bx-python@0.7.3 --export > /export/galaxy-central/database/files/jobdepsvgwf9D3d0352aa83162d618d7e804136167d40809c1ceb7f272dde335f1a50fa2956a8/__bx-python@0.7.3
galaxy.tools.deps.conda_util DEBUG 2017-02-17 22:55:50,340 Executing command: /export/tool_deps/_conda/bin/conda create -y --unknown --offline --prefix /export/galaxy-central/database/job_working_directory/000/42/conda-env --file /export/galaxy-central/database/files/jobdepsvgwf9D3d0352aa83162d618d7e804136167d40809c1ceb7f272dde335f1a50fa2956a8/__bx-python@0.7.3 > /dev/null
galaxy.tools.deps.conda_util DEBUG 2017-02-17 22:55:51,543 Executing command: /export/tool_deps/_conda/bin/conda clean --tarballs -y
galaxy.tools.deps.conda_util DEBUG 2017-02-17 22:55:51,904 Executing command: /export/tool_deps/_conda/bin/conda list --name __six@1.9.0 --export > /export/galaxy-central/database/files/jobdeps2RMTFW1d16fd222aee9584c428e1c21f2685ccf162660faa74d7ed71ee64fcc5a332dd/__six@1.9.0
galaxy.tools.deps.conda_util DEBUG 2017-02-17 22:55:52,279 Executing command: /export/tool_deps/_conda/bin/conda install -y --unknown --offline --prefix /export/galaxy-central/database/job_working_directory/000/42/conda-env --file /export/galaxy-central/database/files/jobdeps2RMTFW1d16fd222aee9584c428e1c21f2685ccf162660faa74d7ed71ee64fcc5a332dd/__six@1.9.0 > /dev/null
galaxy.tools.deps.conda_util DEBUG 2017-02-17 22:55:52,969 Executing command: /export/tool_deps/_conda/bin/conda clean --tarballs -y
galaxy.jobs.command_factory INFO 2017-02-17 22:55:53,366 Built script [/export/galaxy-central/database/job_working_directory/000/42/tool_script.sh] for tool command [[ "$CONDA_DEFAULT_ENV" = "/export/galaxy-central/database/job_working_directory/000/42/conda-env" ] || . /export/tool_deps/_conda/bin/activate '/export/galaxy-central/database/job_working_directory/000/42/conda-env' > conda_activate.log 2>&1 ; [ "$CONDA_DEFAULT_ENV" = "/export/galaxy-central/database/job_working_director
y/000/42/conda-env" ] || . /export/tool_deps/_conda/bin/activate '/export/galaxy-central/database/job_working_directory/000/42/conda-env' > conda_activate.log 2>&1 ; PACKAGE_BASE=/export/tool_deps/cufflinks/2.2.1/devteam/cufflinks/83bec71c5c9f; export PACKAGE_BASE; . /export/tool_deps/cufflinks/2.2.1/devteam/cufflinks/83bec71c5c9f/env.sh; cufflinks 2>&1 | head -n 1 > /export/galaxy-central/database/files/GALAXY_VERSION_STRING_42 2>&1; python '/shed_tools/toolshed.g2.bx.psu.edu/repos/
devteam/cufflinks/83bec71c5c9f/cufflinks/cufflinks_wrapper.py' --input '/export/galaxy-central/database/files/000/dataset_13.dat' --assembled-isoforms-output '/export/galaxy-central/database/files/000/dataset_170.dat' --num-threads "${GALAXY_SLOTS:-4}" -I 300000 -F 0.1 -j 0.15]
galaxy.tools.deps DEBUG 2017-02-17 22:55:53,694 Find dependency samtools version None
galaxy.tools.deps DEBUG 2017-02-17 22:55:53,694 Dependency samtools not found.
galaxy.jobs.runners DEBUG 2017-02-17 22:55:53,699 (42) command is: mkdir -p working; cd working; /export/galaxy-central/database/job_working_directory/000/42/tool_script.sh; return_code=$?; if [ -f /export/galaxy-central/database/job_working_directory/000/42/working/genes.fpkm_tracking ] ; then cp /export/galaxy-central/database/job_working_directory/000/42/working/genes.fpkm_tracking /export/galaxy-central/database/files/000/dataset_168.dat ; fi; if [ -f /export/galaxy-central/datab
ase/job_working_directory/000/42/working/isoforms.fpkm_tracking ] ; then cp /export/galaxy-central/database/job_working_directory/000/42/working/isoforms.fpkm_tracking /export/galaxy-central/database/files/000/dataset_169.dat ; fi; if [ -f /export/galaxy-central/database/job_working_directory/000/42/working/global_model.txt ] ; then cp /export/galaxy-central/database/job_working_directory/000/42/working/global_model.txt /export/galaxy-central/database/files/000/dataset_171.dat ; fi;
if [ -f /export/galaxy-central/database/job_working_directory/000/42/working/skipped.gtf ] ; then cp /export/galaxy-central/database/job_working_directory/000/42/working/skipped.gtf /export/galaxy-central/database/files/000/dataset_172.dat ; fi; cd '/export/galaxy-central/database/job_working_directory/000/42'; python "/export/galaxy-central/database/job_working_directory/000/42/set_metadata__zxLnB.py" "/export/galaxy-central/database/files/tmp6MwK_r" "/export/galaxy-central/database
/job_working_directory/000/42/working/galaxy.json" "/export/galaxy-central/database/job_working_directory/000/42/metadata_in_HistoryDatasetAssociation_168_HuXiIp,/export/galaxy-central/database/job_working_directory/000/42/metadata_kwds_HistoryDatasetAssociation_168_lnoFl6,/export/galaxy-central/database/job_working_directory/000/42/metadata_out_HistoryDatasetAssociation_168_UgW8gx,/export/galaxy-central/database/job_working_directory/000/42/metadata_results_HistoryDatasetAssociation
_168_LoKRRc,/export/galaxy-central/database/files/000/dataset_168.dat,/export/galaxy-central/database/job_working_directory/000/42/metadata_override_HistoryDatasetAssociation_168_JXF7mI" "/export/galaxy-central/database/job_working_directory/000/42/metadata_in_HistoryDatasetAssociation_169_lDidRr,/export/galaxy-central/database/job_working_directory/000/42/metadata_kwds_HistoryDatasetAssociation_169_Hf7Gpy,/export/galaxy-central/database/job_working_directory/000/42/metadata_out_Hist
oryDatasetAssociation_169_g84KPL,/export/galaxy-central/database/job_working_directory/000/42/metadata_results_HistoryDatasetAssociation_169_5DNNZm,/export/galaxy-central/database/files/000/dataset_169.dat,/export/galaxy-central/database/job_working_directory/000/42/metadata_override_HistoryDatasetAssociation_169_F_DY8Y" "/export/galaxy-central/database/job_working_directory/000/42/metadata_in_HistoryDatasetAssociation_170_t92mV9,/export/galaxy-central/database/job_working_directory/
000/42/metadata_kwds_HistoryDatasetAssociation_170_pdQ67L,/export/galaxy-central/database/job_working_directory/000/42/metadata_out_HistoryDatasetAssociation_170_riltIN,/export/galaxy-central/database/job_working_directory/000/42/metadata_results_HistoryDatasetAssociation_170_sdtBLu,/export/galaxy-central/database/files/000/dataset_170.dat,/export/galaxy-central/database/job_working_directory/000/42/metadata_override_HistoryDatasetAssociation_170_weK9U7" "/export/galaxy-central/datab
ase/job_working_directory/000/42/metadata_in_HistoryDatasetAssociation_171_f5qeSO,/export/galaxy-central/database/job_working_directory/000/42/metadata_kwds_HistoryDatasetAssociation_171_Y7CAuA,/export/galaxy-central/database/job_working_directory/000/42/metadata_out_HistoryDatasetAssociation_171_foy9fQ,/export/galaxy-central/database/job_working_directory/000/42/metadata_results_HistoryDatasetAssociation_171_PZ3Tb0,/export/galaxy-central/database/files/000/dataset_171.dat,/export/ga
laxy-central/database/job_working_directory/000/42/metadata_override_HistoryDatasetAssociation_171_8L0huR" "/export/galaxy-central/database/job_working_directory/000/42/metadata_in_HistoryDatasetAssociation_172_Yzt_9c,/export/galaxy-central/database/job_working_directory/000/42/metadata_kwds_HistoryDatasetAssociation_172_d7UMDO,/export/galaxy-central/database/job_working_directory/000/42/metadata_out_HistoryDatasetAssociation_172_tICeWh,/export/galaxy-central/database/job_working_dir
ectory/000/42/metadata_results_HistoryDatasetAssociation_172_ydyV7j,/export/galaxy-central/database/files/000/dataset_172.dat,/export/galaxy-central/database/job_working_directory/000/42/metadata_override_HistoryDatasetAssociation_172_xus4iI" 5242880; sh -c "exit $return_code"
galaxy.jobs.runners.drmaa DEBUG 2017-02-17 22:55:53,735 (42) submitting file /export/galaxy-central/database/job_working_directory/000/42/galaxy_42.sh
galaxy.jobs.runners.drmaa DEBUG 2017-02-17 22:55:53,736 (42) native specification is: --ntasks=1 --share
galaxy.jobs.runners.drmaa INFO 2017-02-17 22:55:53,739 (42) queued as 42
galaxy.jobs DEBUG 2017-02-17 22:55:53,739 (42) Persisting job destination (destination id: slurm_cluster)
galaxy.jobs.runners.drmaa DEBUG 2017-02-17 22:55:53,910 (42/42) state change: job is running
galaxy.jobs.runners.drmaa DEBUG 2017-02-17 22:55:58,042 (42/42) state change: job finished normally
galaxy.jobs.output_checker INFO 2017-02-17 22:55:58,092 Job 42: Fatal error: Exit code 1 ()
galaxy.jobs.output_checker DEBUG 2017-02-17 22:55:58,092 Tool exit code indicates an error, failing job.
galaxy.jobs DEBUG 2017-02-17 22:55:58,134 (42) setting dataset 168 state to ERROR
galaxy.jobs DEBUG 2017-02-17 22:55:58,151 (42) setting dataset 169 state to ERROR
galaxy.jobs DEBUG 2017-02-17 22:55:58,168 (42) setting dataset 170 state to ERROR
galaxy.jobs DEBUG 2017-02-17 22:55:58,186 (42) setting dataset 171 state to ERROR
galaxy.jobs DEBUG 2017-02-17 22:55:58,202 (42) setting dataset 172 state to ERROR
galaxy.jobs INFO 2017-02-17 22:55:58,294 Collecting metrics for Job 42
galaxy.jobs DEBUG 2017-02-17 22:55:58,306 job 42 ended (finish() executed in (219.516 ms))

```

**Comment by nsoranzo** [comment]
```
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,979 Find dependency bx-python version 0.7.3
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,979 Using dependency bx-python version 0.7.3 of type conda
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,979 Find dependency six version 1.9.0 galaxy.tools.deps DEBUG 2017-02-17 22:55:49,980 Using dependency six version 1.9.0 of type conda
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,980 Find dependency cufflinks version 2.2.1
galaxy.tools.deps DEBUG 2017-02-17 22:55:49,982 Using dependency cufflinks version 2.2.1 of type tool_shed_package
```

Here's the issue, you have a mix of Conda and TS packages, try to uninstall package_cufflinks_2_2_1 and the corresponding dependency in the cufflinks package. 

@lecorguille @mvdbeek We need to drop tool_dependencies.xml from the cufflinks tool collection.

**Comment by tiagoantao** [comment]
I will try to do this, but my point here is that this was done on a fresh installation following the defaults, using the most recent packages from the toolshed. I thus suspect that the default behavior is broke.

**Comment by mvdbeek** [comment]
> @lecorguille @mvdbeek We need to drop tool_dependencies.xml from the cufflinks tool collection.

That's done and merged in #453. I have removed the import from galaxy and copied the two helper functions it was importing directly into the wrapper, so cufflinks doesn't depend on bx-python or six anymore. I didn't test it with python3, but it should work.

**Comment by nsoranzo** [comment]
I think that's fixed, feel free to reopen if not.
